pkg/controllers/vdb/altersubcluster_reconciler.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	vmeta "github.com/vertica/vertica-kubernetes/pkg/meta"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/altersc"
+	"github.com/vertica/vertica-kubernetes/pkg/vk8s"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// AlterSubclusterTypeReconciler will change a subcluster type
+type AlterSubclusterTypeReconciler struct {
+	VRec       *VerticaDBReconciler
+	Log        logr.Logger
+	Vdb        *vapi.VerticaDB // Vdb is the CRD we are acting on.
+	PFacts     *PodFacts
+	Dispatcher vadmin.Dispatcher
+}
+
+// MakeAlterSubclusterTypeReconciler will build a AlterSubclusterTypeReconciler object
+func MakeAlterSubclusterTypeReconciler(vdbrecon *VerticaDBReconciler, log logr.Logger,
+	vdb *vapi.VerticaDB, pfacts *PodFacts, dispatcher vadmin.Dispatcher) controllers.ReconcileActor {
+	return &AlterSubclusterTypeReconciler{
+		VRec:       vdbrecon,
+		Log:        log.WithName("AlterSubclusterTypeReconciler"),
+		Vdb:        vdb,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+	}
+}
+
+func (a *AlterSubclusterTypeReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	if a.PFacts.SandboxName == vapi.MainCluster {
+		return ctrl.Result{}, nil
+	}
+
+	if err := a.PFacts.Collect(ctx, a.Vdb); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	scs, err := a.findSandboxSubclustersToAlter()
+	if err != nil || len(scs) == 0 {
+		return ctrl.Result{}, err
+	}
+
+	return a.alterSubclusters(ctx, scs)
+}
+
+// findSandboxSubclustersToAlter returns a list of subclusters whose type needs to be changed
+func (a *AlterSubclusterTypeReconciler) findSandboxSubclustersToAlter() ([]*vapi.Subcluster, error) {
+	scs := []*vapi.Subcluster{}
+	sb := a.Vdb.GetSandbox(a.PFacts.SandboxName)
+	if sb == nil {
+		return scs, fmt.Errorf("could not find sandbox %s", a.PFacts.SandboxName)
+	}
+	scMap := a.Vdb.GenSubclusterMap()
+	for i := range sb.Subclusters {
+		sc := scMap[sb.Subclusters[i].Name]
+		if sc == nil {
+			return scs, fmt.Errorf("could not find subcluster %s", sb.Subclusters[i].Name)
+		}
+		targetType, found := sc.Annotations[vmeta.ParentSubclusterTypeAnnotation]
+		if found && targetType == vapi.PrimarySubcluster && !sc.IsPrimary() {
+			scs = append(scs, sc)
+		}
+	}
+	return scs, nil
+}
+
+func (a *AlterSubclusterTypeReconciler) alterSubclusters(ctx context.Context, scs []*vapi.Subcluster) (ctrl.Result, error) {
+	for _, sc := range scs {
+		_, err := vk8s.UpdateVDBWithRetry(ctx, a.VRec, a.Vdb, func() (bool, error) {
+			scMap := a.Vdb.GenSubclusterMap()
+			vdbSc, found := scMap[sc.Name]
+			if !found {
+				return false, fmt.Errorf("subcluster %q missing in vdb %q", sc.Name, a.Vdb.Name)
+			}
+			vdbSc.Type = vapi.SandboxPrimarySubcluster
+			return true, nil
+		})
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		res, err := a.alterSubclusterType(ctx, sc)
```

Comment:
Move line 108-112 above line 96. We should update vdb after vertica side change is done.